packages/mobile/src/components/audio/Audio.tsx

```diff
@@ -349,211 +385,162 @@ export const Audio = () => {
     setListenLoggedForTrack(false)
   }, [counter, resetPositionForSameTrack])
 
-  const { loading: loadingOfflineTrack, value: offlineTrackUri } =
-    useOfflineTrackUri(track?.track_id.toString())
-  const { loading: loadingNextOfflineTrack, value: nextOfflineTrackUri } =
-    useOfflineTrackUri(nextTrack?.track_id.toString())
-
-  const streamingUri = useMemo(() => {
-    return track && isReachable
-      ? apiClient.makeUrl(`/tracks/${encodeHashId(track.track_id)}/stream`)
-      : null
-  }, [isReachable, track])
-  const nextStreamingUri = useMemo(() => {
-    return nextTrack && isReachable
-      ? apiClient.makeUrl(`/tracks/${encodeHashId(nextTrack.track_id)}/stream`)
-      : null
-  }, [isReachable, nextTrack])
-
-  const gateways = trackOwner
-    ? audiusBackendInstance.getCreatorNodeIPFSGateways(
-        trackOwner.creator_node_endpoint
-      )
-    : []
-  const nextGateways = nextTrackOwner
-    ? audiusBackendInstance.getCreatorNodeIPFSGateways(
-        nextTrackOwner.creator_node_endpoint
+  // Ref to keep track of the queue in the track player vs the queue in state
+  const queueListRef = useRef<string[]>([])
+  // Ref to ensure that we do not try to update while we are already updating
+  const updatingQueueRef = useRef<boolean>(false)
+
+  const handleQueueChange = useCallback(async () => {
+    const refUids = queueListRef.current
+    const trackUids = queueShuffle ? queueShuffleTrackUids : queueTrackUids
+    const tracks = queueShuffle ? queueShuffleTracks : queueTracks
+
+    if (updatingQueueRef.current || isEqual(refUids, trackUids)) return
+    updatingQueueRef.current = true
+    // Check if this is a new queue or we are appending to the queue
+    const isQueueAppend =
+      refUids.length > 0 && isEqual(trackUids.slice(0, refUids.length), refUids)
+    const newQueueTracks = isQueueAppend ? tracks.slice(refUids.length) : tracks
+
+    const localTrackImageSourceMap = {}
```

Comment:
Instead of computing these two maps first, could we just make it part of the `newTrackData` computation?

I assume you wanted to separate it because of the async calls, but maybe we could use a for loop to iterate newQueueTracks when computing `newTrackData`. Or to avoid a for loop we could use promise.all:

```ts
await Promise.all(newQueueTracks.map(async (track) => {
  // stuff
}));
```